com.unity.shadergraph/Editor/Data/Graphs/CategoryData.cs

```diff
@@ -17,37 +17,117 @@ public string name
         }
+        public string categoryGuid => this.objectId;
 
         [SerializeField]
-        Guid m_CategoryGuid;
+        List<JsonRef<ShaderInput>> m_ChildObjectList = new List<JsonRef<ShaderInput>>();
 
-        public Guid categoryGuid
+        public RefValueEnumerable<ShaderInput> Children => m_ChildObjectList.SelectValue();
+
+        // We expose the object list as a HashSet of their objectIDs for faster existence checks
+        HashSet<string> m_ChildObjectIDSet = new HashSet<string>();
+
+        public int childCount => m_ChildObjectIDSet.Count;
+
+        public void InsertItemIntoCategory(ShaderInput itemToAdd, int insertionIndex = -1)
         {
-            get => m_CategoryGuid;
-            set => m_CategoryGuid = value;
+            if (itemToAdd == null)
+            {
+                AssertHelpers.Fail("Tried to insert item into category that was null.");
+                return;
+            }
+
+            if (insertionIndex == -1)
+            {
+                m_ChildObjectList.Add(itemToAdd);
+                m_ChildObjectIDSet.Add(itemToAdd.objectId);
+            }
+            else
+            {
+                m_ChildObjectList.Insert(insertionIndex, itemToAdd);
+                // Recreate the hash-set as the data structure does not allow for moving/inserting elements
+                m_ChildObjectIDSet.Clear();
+                foreach (var childObject in m_ChildObjectList)
+                {
+                    m_ChildObjectIDSet.Add(childObject.value.objectId);
+                }
+            }
         }
 
-        // We store Guids as a list out to the graph asset
-        [SerializeField]
-        List<Guid> m_ChildItemIDList;
+        public void RemoveItemFromCategory(ShaderInput itemToRemove)
+        {
+            m_ChildObjectList.Remove(itemToRemove);
+            if (m_ChildObjectIDSet.Contains(itemToRemove.objectId))
+                m_ChildObjectIDSet.Remove(itemToRemove.objectId);
+        }
+
+        public void MoveItemInCategory(ShaderInput itemToMove, int newIndex)
+        {
+            m_ChildObjectList.Remove(itemToMove);
+            m_ChildObjectList.Insert(newIndex,itemToMove);
+            // Recreate the hash-set as the data structure does not allow for moving/inserting elements
+            m_ChildObjectIDSet.Clear();
+            foreach (var childObject in m_ChildObjectList)
+            {
+                m_ChildObjectIDSet.Add(childObject.value.objectId);
+            }
+        }
+
+        public bool IsItemInCategory(ShaderInput itemToCheck)
+        {
+            return m_ChildObjectIDSet.Contains(itemToCheck.objectId);
+        }
 
-        HashSet<Guid> m_ChildItemIDSet;
-        // We expose Guids as a HashSet for faster existence checks
-        public HashSet<Guid> childItemIDSet
+        public bool IsNamedCategory()
         {
-            get => m_ChildItemIDSet;
-            set => m_ChildItemIDSet = value;
+            return name != String.Empty;
```

Comment:
Is null possible here? Would it be better to use `string.IsNullOrEmpty`?

Reply:
The category names are initialized to `String.Empty` at the start so I believe nulls are not possible here, which would also be an assumption I believe a lot of things in the code depend on